### PR TITLE
Updating to APIv2

### PR DESCRIPTION
**Pull request by NaliniSewak**
Updating android quickstart sample code to APIv2. Handlers are updated to use new fromJson() object builders.

All code snippets were formatted according to google-java-format.

### CODE REVIEW

**Comment by NaliniSewak** [comment]
Linked nsewak@google.com to Github Account.

File: app/build.gradle [hunk 1]
```diff
@@ -24,7 +24,8 @@ android {
 dependencies {
     implementation fileTree(dir: 'libs', include: ['*.jar'])
 
-    implementation 'com.google.android.gms:play-services:11.8.0'
```
Comment (fstanis) [inline app/build.gradle]: If you're replacing `play-services` with `play-services-wallet`, please update or remove the comment below as well.

Also note that we had appcompat in the past, but this caused problems with some partners who copied it from the sample, resulting in having the wrong version. I advise against including.

File: app/src/main/java/com/google/android/gms/samples/wallet/PaymentsUtil.java [hunk 2]
```diff
@@ -53,129 +55,227 @@ private PaymentsUtil() {
...
+                "parameters",
+                new JSONObject()
+                        .put("protocolVersion", "ECv1")
                         // Omitting the publicKey will result in a request for unencrypted data.
-                        // Please refer to the documentation for more information on unencrypted
-                        // requests.
-                        .addParameter("publicKey", Constants.DIRECT_TOKENIZATION_PUBLIC_KEY)
-                        .build();
+                        // Please refer to the documentation for more information on unencrypted requests.
+                        .put("publicKey", Constants.DIRECT_TOKENIZATION_PUBLIC_KEY));
```
Comment (niallkennedy) [inline app/src/main/java/com/google/android/gms/samples/wallet/PaymentsUtil.java]: check if DIRECT_TOKENIZATION_PUBLIC_KEY is set or is default at start of method. reject early

File: app/src/main/java/com/google/android/gms/samples/wallet/PaymentsUtil.java [hunk 3]
```diff
@@ -53,129 +55,227 @@ private PaymentsUtil() {
...
-                        .addParameter("gateway", Constants.GATEWAY_TOKENIZATION_NAME);
-        for (Pair<String, String> param : Constants.GATEWAY_TOKENIZATION_PARAMETERS) {
-            paramsBuilder.addParameter(param.first, param.second);
-        }
+    private static JSONObject getGatewayTokenizationSpecification() throws JSONException {
+        JSONObject tokenizationSpecification = new JSONObject();
 
-        return createPaymentDataRequest(transactionInfo, paramsBuilder.build());
+        tokenizationSpecification.put("type", "PAYMENT_GATEWAY");
+        JSONObject parameters = new JSONObject(Constants.GATEWAY_TOKENIZATION_PARAMETERS);
```
Comment (niallkennedy) [inline app/src/main/java/com/google/android/gms/samples/wallet/PaymentsUtil.java]: reject early if GATEWAY_TOKENIZATION_PARAMETERS is empty

File: app/src/main/java/com/google/android/gms/samples/wallet/PaymentsUtil.java [hunk 4]
```diff
@@ -53,129 +55,227 @@ private PaymentsUtil() {
...
-                        .setPaymentMethodTokenizationType(
-                                WalletConstants.PAYMENT_METHOD_TOKENIZATION_TYPE_DIRECT)
+    private static JSONObject getDirectTokenizationSpecification() throws JSONException {
 
+        JSONObject tokenizationSpecification = new JSONObject();
+        tokenizationSpecification.put("type", "DIRECT");
+        tokenizationSpecification.put(
+                "parameters",
+                new JSONObject()
+                        .put("protocolVersion", "ECv1")
```
Comment (niallkennedy) [inline app/src/main/java/com/google/android/gms/samples/wallet/PaymentsUtil.java]: add direct tokenization protocol version to Constants

File: app/src/main/java/com/google/android/gms/samples/wallet/PaymentsUtil.java [hunk 5]
```diff
@@ -53,129 +55,227 @@ private PaymentsUtil() {
...
+        cardPaymentMethod.put("type", "CARD");
+
+        JSONObject parameters = new JSONObject();
+        parameters.put("allowedAuthMethods", getAllowedCardAuthMethods());
+        parameters.put("allowedCardNetworks", getAllowedCardNetworks());
+        // Optionally, you can add billing address/phone number associated with a CARD payment method.
+        parameters.put("billingAddressRequired", true);
 
-        return createPaymentDataRequest(transactionInfo, params);
+        JSONObject billingAddrParams = new JSONObject();
```
Comment (niallkennedy) [inline app/src/main/java/com/google/android/gms/samples/wallet/PaymentsUtil.java]: billingAddressParameters

File: app/src/main/java/com/google/android/gms/samples/wallet/PaymentsUtil.java [hunk 6]
```diff
@@ -53,129 +55,227 @@ private PaymentsUtil() {
...
+        try {
+            JSONObject isReadyToPayRequest = getBaseRequest();
+            isReadyToPayRequest.put(
+                    "allowedPaymentMethods", new JSONArray().put(getBaseCardPaymentMethod()));
+
+            // An optional shipping address requirement is a top-level property of the PaymentDataRequest
+            // JSON object. Omitting all together means all countries are supported.
+            isReadyToPayRequest.put("shippingAddressRequired", true);
+
+            JSONObject shippingAddrParams = new JSONObject();
```
Comment (niallkennedy) [inline app/src/main/java/com/google/android/gms/samples/wallet/PaymentsUtil.java]: shippingAddressParameters

File: app/src/main/java/com/google/android/gms/samples/wallet/PaymentsUtil.java [hunk 7]
```diff
@@ -53,129 +55,227 @@ private PaymentsUtil() {
...
-        IsReadyToPayRequest.Builder request = IsReadyToPayRequest.newBuilder();
-        for (Integer allowedMethod : Constants.SUPPORTED_METHODS) {
-            request.addAllowedPaymentMethod(allowedMethod);
+    public static Optional<JSONObject> getIsReadyToPayRequest() {
+        try {
+            JSONObject isReadyToPayRequest = getBaseRequest();
+            isReadyToPayRequest.put(
+                    "allowedPaymentMethods", new JSONArray().put(getBaseCardPaymentMethod()));
+
+            // An optional shipping address requirement is a top-level property of the PaymentDataRequest
```
Comment (niallkennedy) [inline app/src/main/java/com/google/android/gms/samples/wallet/PaymentsUtil.java]: shipping fields should not be included in the IsReadyToPayRequest

File: app/src/main/java/com/google/android/gms/samples/wallet/PaymentsUtil.java [hunk 8]
```diff
@@ -53,129 +55,227 @@ private PaymentsUtil() {
...
+     * Card networks supported by your app and your gateway.
+     *
+     * <p>TODO: Confirm card networks supported by your app and gateway & update in Constants.java.
+     *
+     * @return Allowed card networks
+     * @see <a
+     *     href="https://developers.google.com/pay/api/android/reference/object#CardParameters">CardParameters</a>
+     */
+    private static JSONArray getAllowedCardNetworks() {
+        JSONArray allowedCardNetworks = new JSONArray();
```
Comment (niallkennedy) [inline app/src/main/java/com/google/android/gms/samples/wallet/PaymentsUtil.java]: why not return a JSONArray constructed from the List?
https://developer.android.com/reference/org/json/JSONArray.html#JSONArray(java.util.Collection)

File: app/build.gradle [hunk 9]
```diff
@@ -24,10 +25,11 @@ android {
 dependencies {
     implementation fileTree(dir: 'libs', include: ['*.jar'])
 
+    /* "play-services-wallet" is the only module needed for the Google Pay API and this sample app
+     but the "play-services" module can also be used.
+     The version provided is the minimum version required for the most recent version of the
```
Comment (niallkennedy) [inline app/build.gradle]: most recent major version

File: app/build.gradle [hunk 10]
```diff
@@ -24,10 +25,11 @@ android {
...
+     Google Pay API library. Your app may choose a dependency greater than the value above.
```
Comment (niallkennedy) [inline app/build.gradle]: the value is below this line of text: line 34 vs. line 31

File: app/src/main/java/com/google/android/gms/samples/wallet/CheckoutActivity.java [hunk 11]
```diff
@@ -190,52 +194,45 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
...
-                        tokenObject
-                                .getJSONObject("paymentMethodData")
-                                .getJSONObject("tokenizationData")
-                                .getString("token"));
-            } catch (JSONException e) {
-                Log.e("handlePaymentSuccess failed", e.toString());
-                return;
+        try {
+            paymentMethodData = new JSONObject(paymentInformation).getJSONObject("paymentMethodData");
+            // If the gateway is set to example, no payment information is returned - instead, the
```
Comment (niallkennedy) [inline app/src/main/java/com/google/android/gms/samples/wallet/CheckoutActivity.java]: "example" to match the quoted style of response value

File: app/src/main/java/com/google/android/gms/samples/wallet/CheckoutActivity.java [hunk 12]
```diff
@@ -190,52 +194,45 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
...
+            // token will only consist of "examplePaymentMethodToken".
+            if (paymentMethodData.getJSONObject("tokenizationData").getString("type").equals("PAYMENT_GATEWAY")
```
Comment (niallkennedy) [inline app/src/main/java/com/google/android/gms/samples/wallet/CheckoutActivity.java]: consider reformatting with consistent newlines and indentation to make each paymentMethodData getter and comparison easier to read

if (paymentMethodData
        .getJSONObject("tokenizationData")
        .getString("type")
        .equals("PAYMENT_GATEWAY")
    && paymentMethodData
        .getJSONObject("tokenizationData")
        .getString("token")
        .equals("examplePaymentMethodToken")) {}

File: app/src/main/java/com/google/android/gms/samples/wallet/Constants.java [hunk 13]
```diff
@@ -21,36 +21,47 @@
...
-    //     GATEWAY_TOKENIZATION_NAME and GATEWAY_TOKENIZATION_PARAMETERS per the instructions they
-    //     provided. You don't need to update DIRECT_TOKENIZATION_PUBLIC_KEY.
-    // 5.  If you're using direct integration, please consult the documentation to learn about
-    //     next steps.
-
-    // Changing this to ENVIRONMENT_PRODUCTION will make the API return real card information.
-    // Please refer to the documentation to read about the required steps needed to enable
-    // ENVIRONMENT_PRODUCTION.
+    /**
+     * Changing this to ENVIRONMENT_PRODUCTION will make the API return real card information.
```
Comment (niallkennedy) [inline app/src/main/java/com/google/android/gms/samples/wallet/Constants.java]: chargeable card information

File: app/src/main/java/com/google/android/gms/samples/wallet/PaymentsUtil.java [hunk 14]
```diff
@@ -268,6 +254,18 @@ private static JSONObject getMerchantInfo() throws JSONException {
                     "allowedPaymentMethods", new JSONArray().put(PaymentsUtil.getCardPaymentMethod()));
             paymentDataRequest.put("transactionInfo", PaymentsUtil.getTransactionInfo(price));
             paymentDataRequest.put("merchantInfo", PaymentsUtil.getMerchantInfo());
+
+            // An optional shipping address requirement is a top-level property of the PaymentDataRequest
+            // JSON object. Omitting all together means all countries are supported.
```
Comment (niallkennedy) [inline app/src/main/java/com/google/android/gms/samples/wallet/PaymentsUtil.java]: remove "omitting all together" sentence. it is incorrect, as country restrictions is in a different property

File: app/src/main/java/com/google/android/gms/samples/wallet/PaymentsUtil.java [hunk 15]
```diff
@@ -34,250 +34,251 @@
...
+   * @throws JSONException
+   * @see <a href=
+   *     "https://developers.google.com/pay/api/android/reference/object#PaymentMethodTokenizationSpecification">PaymentMethodTokenizationSpecification</a>
+   */
+  private static JSONObject getGatewayTokenizationSpecification()
+      throws JSONException, RuntimeException {
+    if (Constants.PAYMENT_GATEWAY_TOKENIZATION_PARAMETERS.isEmpty()) {
+      throw new RuntimeException(
+          "Please edit the Constants.java file to add gateway name and other parameters your"
+              + "processor requires");
```
Comment (niallkennedy) [inline app/src/main/java/com/google/android/gms/samples/wallet/PaymentsUtil.java]: add space to be displayed in the final string between "your" and "processor"

File: app/src/main/java/com/google/android/gms/samples/wallet/CheckoutActivity.java [hunk 16]
```diff
@@ -38,254 +38,258 @@
...
-                        break;
-                    default:
-                        // Do nothing.
-                }
-
-                // Re-enables the Pay with Google button.
-                mGooglePayButton.setClickable(true);
-                break;
-        }
+        // Re-enables the Pay with Google button.
```
Comment (niallkennedy) [inline app/src/main/java/com/google/android/gms/samples/wallet/CheckoutActivity.java]: Google Pay payment button

**Pull request merged**